accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -579,6 +617,153 @@ all occurrences of `S*` in `T` shall be replaced with `S`.  As a result, legacy
+the class) has all generic arguments (and all sub-components of the generic
+arguments) marked as legacy types.
+
+If a class `C` in a legacy library implements the same generic class `I` more
+than once, it is an error if the `LEGACY_ERASURE` of all such super-interfaces
+are not all syntactically equal.  For the purposes of runtime subtyping checks,
+`C` is considered to implement the canonical `LEGACY_ERASURE` of the
+super-interfaces in question.
+
+A member which is defined in a class in a legacy library (whether concrete or
+abstract), is given a signature in which every type is a legacy type.  It is an
+error if the signature of a member is not a correct override of all members of
+the same name in super-interfaces of the class, using the legacy subtyping
+rules.
+
+#### Classes defined in legacy libraries as seen from opted-in libraries
+
+The `LEGACY_TOP_MERGE` of two types `T` and `S` is the unique type `R` defined
+as:
+ - `LEGACY_TOP_MERGE(T?, S?) = LEGACY_TOP_MERGE(T, S)?`
+ - `LEGACY_TOP_MERGE(T?, S*) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S?) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S*) = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T*, S)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T, S*)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T?, S)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(T, S?)  = LEGACY_TOP_MERGE(T, S)*`
+ - `LEGACY_TOP_MERGE(Never, Null)  = Null`
+ - `LEGACY_TOP_MERGE(Null, Never)  = Null`
+ - `LEGACY_TOP_MERGE(void, void)  = void`
+ - `LEGACY_TOP_MERGE(Object, Object)  = Object`
+ - `LEGACY_TOP_MERGE(dynamic, dynamic)  = Object`
+ - `LEGACY_TOP_MERGE(Object, void)  = void`
+   - And the reverse
+ - `LEGACY_TOP_MERGE(dynamic, void)  = void`
+   - And the reverse
+ - `LEGACY_TOP_MERGE(Object, dynamic)  = Object`
+   - And the reverse
+ - And for all other types, recursively applying the transformation over the
+   structure of the type
+   - e.g. `LEGACY_TOP_MERGE(C<T>, C<S>)  = C<LEGACY_TOP_MERGE(T, S)>`
+
+In other words, `LEGACY_TOP_MERGE` takes two types which are structurally equal
+except for the placement of `?` and `*`, the use of `Null` vs `Never`, and the
+particular choice of top types, and finds a single canonical type to represent
+them. The `LEGACY_TOP_MERGE` of two types is not always defined.
+
+If a class which is defined in a legacy library inherits a member with the same
+name from multiple super-interfaces, then error checking is done as usual using
+the legacy typing rules which ignore nullability.  This means that it is valid
+for a legacy class to inherit the same member signature with contradictory
+nullability information. For the purposes of member lookup within the legacy
+library itself, nullability information is ignored, and so it is valid to simply
+erase the nullability information within the legacy library. When referenced
+from an opted-in library, if there are multiple signatures `T0, ... Tn` for a
```

Comment:
Does this mean that each class in a legacy library has two sets of interfaces - one to use from the legacy library, and another to use from opted-in libraries? This would be somewhat computationally expensive in analyzer.

Also, the wording is "lookup within the legacy library itself", is "itself" has a special intention here, or we could say "lookup within **any** legacy library"?

Reply:
Fixed the "itself" issue, the intent is the latter.

Reply:
> Does this mean that each class in a legacy library has two sets of interfaces - one to use from the legacy library, and another to use from opted-in libraries? This would be somewhat computationally expensive in analyzer.

This is the crux of the question.  If we ignore the part of LEGACY_TOP_MERGE which deals with the top types, then the answer is no.  You can just use LEGACY_TOP_MERGE of the Si both inside and outside of the legacy library.  In the legacy library, you just ignore the nullability annotations, in an opted-in library you don't.  The top types change that though, because currently I believe that the analyzer just chooses the first compatible signature.  That is, if I inherit a void Function(Object) from my extends clause and a void Function(dynamic) from the implements clause, I believe that both the analyzer and the CFE will just choose the one from the extends clause.  I can choose to specify it this way, and if it simplifies things significantly I will.

Reply:
Looking into implementation it seems that analyzer looks through the list of candidate signature backward (to choose the last override, e.g. from the mixin in `extends S with M`), so will choose the last signature from the list of candidates such that this signature is a valid override of all other candidates. For `class X extends S with M1, M2 implements I1, I2 {}` candidates are from `[I1, I2, S, M1, M2]`.

I'm not entirely sure what we want to achieve here, except of a generic goal to make it opted-in and legacy libraries live together well.

Do we try to solve the problem of "in-out-in", i.e. when opted-it `A` is extended by legacy `B`, and then opted-in `C` extends `B`?

With the approach as in this PR, it looks to me that we try to look at direct opted-in superinterfaces differently from legacy and from opted-in libraries. What if we have two in-between legacy classes? Will we lose information after the first legacy class, because opted-in superinterfaces are not direct anymore?

Just for information, here is a short description what analyzer has implemented for inheritance now. Each class interface has NNBD signatures, regardless whether the class in declared in a legacy or opted-in library. Checking for subtyping of signatures is done by NNBD rules. Signatures declared in legacy libraries have legacy types. Classes declared in legacy libraries will inherit NNBD signatures from superinterfaces, if these interfaces declared in legacy libraries (or themselves inherited these signature from opted-in libraries). But any type arguments to these superinterfaces are specified in a legacy library will be legacy types, so some signatures will be combinations of legacy and NNBD types. When we resolve a reference to a class member, and this reference in a legacy library, we erase nullability from it.

With this approach of using only NNBD signatures, is there still a problem with top types?

Reply:
> Just for information, here is a short description what analyzer has implemented for inheritance now. Each class interface has NNBD signatures, regardless whether the class in declared in a legacy or opted-in library. Checking for subtyping of signatures is done by NNBD rules.

This doesn't really work.  Consider:


opted_in.dart
```
// opted in
class A {
  int? foo(int x) {}
}

class B {
  int foo(int? x) {}
}
```

opted_out.dart
```
// opted out
// @dart = 2.6
import 'opted_in.dart';

class C extends A implements B {}
```

The approach you describe rejects class C there, which we absolutely can't do - we are promising that opting in doesn't break non-opted in code.  So one way or the other, we need to suppress the error in `C`.  That's the point of using LEGACY_SUBTYPE for the override/inheritance checking in legacy libraries.

So that's one issue.  The second issue is now suppose we have the following:

```dart
//opted in
import 'opted_out.dart';
void test() {
  new C().foo(null).isEven; 
} 
```
What signature does `new C().foo` have?  Choices that I see are:
 1) make it legacy
 2) pick the "first one" in some order (current analyzer behavior)
 3) compute a "merged" signature that preserves as much information from the super-interfaces as possible.

I'm basically fine with any of these options, though I think that 2) is the least helpful - it arbitrarily picks between some set of potentially incompatible signatures.

1) is less helpful in the common case that there is not a conflict between the signatures.  If an opted in library uses a method from another opted in library that just happens to be inherited *through* an opted out class, it seems useful to get the proper errors and warnings.  But I don't expect this to be a super common case, so I'm fine with using 1) if it's more convenient.

Reply:
True, for the first example analyzer reports `INVALID_OVERRIDE` because there is no the most specific signature. I remember some discussions that we might consider this a bridge to migrating, but I might be mistaken. Breaking legacy libraries is definitely not something we want to do.

(1) is the most convenient for analyzer. We would erase nullabilities from signatures of super-interfaces (if these methods are declared in opted-in libraries), and then choose the most specific signature. These most specific signatures, actually any signatures that you get from a class declared in a legacy library, would have only legacy types.